test/src/test/java/org/corfudb/integration/ClusterReconfigIT.java

```diff
@@ -599,4 +600,121 @@ public void testSystemDownHandler() throws Exception {
+
+        assertThat(runtime.getLayoutView().getRuntimeLayout()
+                .getLogUnitClient("localhost:9002")
+                .read(1L).get().getAddresses().get(1L).isEmpty()).isFalse();
+
+        shutdownCorfuServer(corfuServer_1);
+        shutdownCorfuServer(corfuServer_2);
+        shutdownCorfuServer(corfuServer_3);
+    }
+
+    /**
+     * Test that a node with a stale layout can recover.
+     * Consider a cluster of 3 nodes - 9000, 9001 and 9002.
+     * 9002 is first shutdown. This node is nowstuck with the layout at epoch 0.
+     * Now 9000 is restarted, forcing an epoch increment.
+     * Next, 9002 is also restarted. The test then asserts that even though 9002 is lagging
+     * by a few epochs, it recovers and is added beck to the chain.
+     */
+    @Test
+    public void healNodesWorkflowTest() throws Exception {
+        // Set up cluster of 3 nodes.
+        final int PORT_0 = 9000;
+        final int PORT_1 = 9001;
+        final int PORT_2 = 9002;
+        Process corfuServer_1 = runUnbootstrappedPersistentServer(corfuSingleNodeHost, PORT_0);
+        Process corfuServer_2 = runUnbootstrappedPersistentServer(corfuSingleNodeHost, PORT_1);
+        Process corfuServer_3 = runUnbootstrappedPersistentServer(corfuSingleNodeHost, PORT_2);
+        final Layout layout = get3NodeLayout();
+
+        final int retries = 3;
+        BootstrapUtil.bootstrap(layout, retries, PARAMETERS.TIMEOUT_SHORT);
+
+        CorfuRuntime runtime = createDefaultRuntime();
+        UUID streamId = CorfuRuntime.getStreamID("testStream");
+        IStreamView stream = runtime.getStreamsView().get(streamId);
+        byte[] payload = "test".getBytes();
+
+        shutdownCorfuServer(corfuServer_3);
+        waitForEpochChange(refreshedEpoch -> refreshedEpoch > layout.getEpoch(), runtime);
+        final Layout layoutAfterFailure1 = runtime.getLayoutView().getLayout();
+
+        shutdownCorfuServer(corfuServer_1);
+        corfuServer_1 = runUnbootstrappedPersistentServer(corfuSingleNodeHost, PORT_0);
+        waitForEpochChange(refreshedEpoch -> refreshedEpoch > layoutAfterFailure1.getEpoch(), runtime);
+        final Layout layoutAfterHeal1 = runtime.getLayoutView().getLayout();
+
+        // Write at address 0.
+        stream.append(payload);
+
+        corfuServer_3 = runUnbootstrappedPersistentServer(corfuSingleNodeHost, PORT_2);
+        waitForEpochChange(refreshedEpoch -> refreshedEpoch > layoutAfterHeal1.getEpoch(), runtime);
+
+        // Write at address 1.
+        stream.append(payload);
+
+        for (int i = 0; i < PARAMETERS.NUM_ITERATIONS_MODERATE; i++) {
+            if (runtime.getLayoutView().getLayout().getUnresponsiveServers().isEmpty()) {
+                break;
+            }
+            runtime.invalidateLayout();
+            Sleep.sleepUninterruptibly(PARAMETERS.TIMEOUT_VERY_SHORT);
+        }
+
+        assertThat(runtime.getLayoutView().getLayout().getUnresponsiveServers()).isEmpty();
```

Comment:
Do we not need to check what nodes are in what segment and if all of them have the data. I know it can be multiple combinations but you can wait for the final expected layout and then check.

Reply:
Done.